synapse/logging/opentracing.py

```diff
@@ -793,48 +793,41 @@ def _tag_args_inner(*args, **kwargs):
     return _tag_args_inner
 
 
-def trace_servlet(servlet_name, extract_context=False):
-    """Decorator which traces a serlet. It starts a span with some servlet specific
-    tags such as the servlet_name and request information
+@contextlib.contextmanager
+def trace_servlet(request: "SynapseRequest", extract_context: bool = False):
+    """Context manager which traces a servlet. It starts a span with some
+    servlet specific tags such as the servlet_name and request information.
 
     Args:
-        servlet_name (str): The name to be used for the span's operation_name
-        extract_context (bool): Whether to attempt to extract the opentracing
+        request
+        extract_context: Whether to attempt to extract the opentracing
             context from the request the servlet is handling.
-
     """
 
-    def _trace_servlet_inner_1(func):
-        if not opentracing:
-            return func
-
-        @wraps(func)
-        async def _trace_servlet_inner(request, *args, **kwargs):
-            request_tags = {
-                "request_id": request.get_request_id(),
-                tags.SPAN_KIND: tags.SPAN_KIND_RPC_SERVER,
-                tags.HTTP_METHOD: request.get_method(),
-                tags.HTTP_URL: request.get_redacted_uri(),
-                tags.PEER_HOST_IPV6: request.getClientIP(),
-            }
-
-            if extract_context:
-                scope = start_active_span_from_request(
-                    request, servlet_name, tags=request_tags
-                )
-            else:
-                scope = start_active_span(servlet_name, tags=request_tags)
-
-            with scope:
-                result = func(request, *args, **kwargs)
+    if opentracing is None:
+        yield
```

Comment:
kinda wondering about the overhead here. There's a `_NullContextManager` in `generic_worker.py` which we could pull out and use...

Reply:
A single if statement per request doesn't fill me with terror, if I'm honest.

Reply:
it's not the if statement I'm worrying about, but more the infrastructure involved in instantiating a generator and constructing a GeneratorContextManager to wrap it.

I'm inclined to agree it's not worth lying awake at night over though.

Reply:
Ah right I see, I forgot that we'd need to generate a context manager object each time